runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactoryTest.java

```diff
@@ -395,6 +400,77 @@ public void doesNotCacheDifferentEnvironments() throws Exception {
+    PortablePipelineOptions portableOptions =
+        PipelineOptionsFactory.as(PortablePipelineOptions.class);
+    portableOptions.setSdkWorkerParallelism(2);
+    portableOptions.setLoadBalanceBundles(true);
+    Struct pipelineOptions = PipelineOptionsTranslation.toProto(portableOptions);
+
+    try (DefaultJobBundleFactory bundleFactory =
+        new DefaultJobBundleFactory(
+            JobInfo.create("testJob", "testJob", "token", pipelineOptions),
+            envFactoryProviderMap,
+            stageIdGenerator,
+            serverInfo)) {
+      OutputReceiverFactory orf = mock(OutputReceiverFactory.class);
+      StateRequestHandler srh = mock(StateRequestHandler.class);
+      when(srh.getCacheTokens()).thenReturn(Collections.emptyList());
+      StageBundleFactory sbf = bundleFactory.forStage(getExecutableStage(environment));
+      RemoteBundle b1 = sbf.getBundle(orf, srh, BundleProgressHandler.ignored());
+      verify(envFactory, Mockito.times(1)).createEnvironment(environment);
+      final RemoteBundle b2 = sbf.getBundle(orf, srh, BundleProgressHandler.ignored());
+      verify(envFactory, Mockito.times(2)).createEnvironment(environment);
+
+      long tms = System.currentTimeMillis();
+      AtomicBoolean closed = new AtomicBoolean();
+      // close to free up environment for another bundle
+      TimerTask closeBundleTask =
+          new TimerTask() {
+            @Override
+            public void run() {
+              try {
+                b2.close();
+                closed.set(true);
+              } catch (Exception e) {
+                throw new RuntimeException(e);
+              }
+            }
+          };
+      new Timer().schedule(closeBundleTask, 100);
+
+      RemoteBundle b3 = sbf.getBundle(orf, srh, BundleProgressHandler.ignored());
+      // ensure we waited for close
+      Assert.assertTrue(System.currentTimeMillis() - tms >= 100 && closed.get());
+
+      verify(envFactory, Mockito.times(2)).createEnvironment(environment);
+      b3.close();
+      b1.close();
+    }
+  }
+
+  @Test
+  public void rejectsStateCachingWithLoadBalancing() throws Exception {
+    PortablePipelineOptions portableOptions =
+        PipelineOptionsFactory.as(PortablePipelineOptions.class);
+    portableOptions.setLoadBalanceBundles(true);
+    ExperimentalOptions options = portableOptions.as(ExperimentalOptions.class);
+    ExperimentalOptions.addExperiment(options, "state_cache_size=1");
+    Struct pipelineOptions = PipelineOptionsTranslation.toProto(options);
+
+    try (DefaultJobBundleFactory bundleFactory =
+        new DefaultJobBundleFactory(
+            JobInfo.create("testJob", "testJob", "token", pipelineOptions),
+            envFactoryProviderMap,
+            stageIdGenerator,
+            serverInfo)) {
+      Assert.fail("state caching should be rejected");
+    } catch (IllegalArgumentException ex) {
+      Assert.assertThat(ex.getMessage(), containsString("state_cache_size"));
+    }
```

Comment:
I like using this syntax because it avoids forgetting to fail after the statement:
```java
Exception e = assertThrows(
    IllegalArgumentException.class, 
    () -> new DefaultJobBundleFactory(
            JobInfo.create("testJob", "testJob", "token", pipelineOptions),
            envFactoryProviderMap,
            stageIdGenerator,
            serverInfo));
assertThat(e.getMessage(), containsString("state_cache_size"));
```